enhancements/20240604-lvm-csi-driver.md

```diff
@@ -0,0 +1,81 @@
+
+### User Experience In Detail
+
+The user creates a VM and starts with the feature with PCI pass-through. The VM is not migratable because the user uses the PCI pass-through feature. Then, the user can create the PVC with the LVM volume to get better performance. The user attaches the PVC to the VM through hotplug and uses the LVM volume in the VM.
+
+### API changes
+TBD
+
+## Design
+
+### Implementation Overview
+
+A overview of the design could be checked below:
+![lvm-csi-driver-arch](20240604-lvm-csi-driver/lvm-csi-driver-arch.png)
+
+### LVM CSI Driver
+At first, we consider two existing solutions: [metal-stack/csi-driver-lvm](https://github.com/metal-stack/csi-driver-lvm) and [topolvm/topolvm](https://github.com/topolvm/topolvm). Both solutions support the CSI topology feature, and other features like volume capability and snapshot are only supported in the topolvm. However, the topolvm is more complex than the csi-driver-lvm. We could try to add the capability and snapshot feature to the csi-driver-lvm and keep it as simple as we want.
+
+So, we chose the [metal-stack/csi-driver-lvm](https://github.com/metal-stack/csi-driver-lvm) as the base LVM CSI driver. This project is lightweight and already has the essential capability for our requirements. We want to introduce it in the first version. Then, we can add more features based on the requirements.
+
+The LVM CSI driver includes the CSI plugin and the LVM provisioner. The CSI plugin handles all CSI requests, while the LVM provisioner manages LVM operations such as creation, deletion, expansion, and snapshotting. This division of responsibilities ensures efficient and effective management of our storage infrastructure.
+
+In the initial version, we propose a significant change that will enhance user flexibility. We plan to allow users to predefine the volume group. The provisioner will then use this predefined volume group and logical volume type (e.g., mirror, stripe) to create the volume. This change will enable users to group similar types of disks together, thereby ensuring optimal performance and capability.
+
+With this change, the StorageClass will now contain two additional parameters: `vgName` and `lvType`. The `vgName` refers to the volume group name, and the `lvType` specifies the logical volume type. These parameters are essential for the LVM provisioner to create the volume according to the user's specifications.
+
+### Node Disk Manager
+Although we can create a volume group manually, the node disk manager can help us do so. The node disk manager will create the volume group based on the disks the user chooses. Then, the user can use the volume group name to create StorageClass and PVC. You can refer https://github.com/harvester/harvester/pull/6015 to find more details about the node disk manager how to create the volume group.
```

Comment:
Does this mean when a user adds a disk to a VG, the VG is created (where should the user specify the VG's attributes)? And the sequential disk adds will extend the VG?

Or should we design a new CRD as a container for disks?

Reply:
From the current behavior (see https://github.com/harvester/node-disk-manager/pull/109/commits/b2080f3179790384bfe806bf7cb3e43c8b7320c3).
If the VG does not exist, we will create the VG. If the VG is already here, we will extend the VG.
For the parameter, I will create a way for people who want to customize the VG.

> Or should we design a new CRD as a container for disks?
I do not think creating a new CRD is a good way because that will make things more complex.
We assume the user knows how VG/LV works. (Or rely on the document).

However, I agree that we should refine the UI flow to allow users to configure various storage backend disks.

Reply:
Can you help clarify if extending the VG involves data movement or unexpected data placement?
Say a user has a VG with two disks and creates a PVC (a stripped LV). Its data are spread into two disks.
If the user later adds a new disk to the VG, will the existing PVC continue spreading data on the first two disks? And do all new PVCs created later spread data on 3 disks?

My point here is when users create PVCs from a storage class, they expect all volumes to have the same characteristics, like availability and performance.

The HEP has to clarify the supported LV types like linear/stripped/RAID/thin provision.

Reply:
> Can you help clarify if extending the VG involves data movement or unexpected data placement? Say a user has a VG with two disks and creates a PVC (a stripped LV). Its data are spread into two disks. If the user later adds a new disk to the VG, will the existing PVC continue spreading data on the first two disks? And do all new PVCs created later spread data on 3 disks?
> 

I just checked with my env, TL;DR, extending VG doesn't involve data movement, so the existing PVC (LV) will only be allocated on the original two PVs.

Reply:
If a new disk is added to the vg (e.q. by `vgextend`), will the parameter of `lvType` change accordingly? 

For example, if there are originally two disks, and add another new disk, will the strip number automatically change from two to three?

Reply:
Thanks @WebberHuang1118,

Hi @bk201 
As @WebberHuang1118  mentioned, the old PVC still uses the number of PV on the creation instead of the new PV count (after the VG is extended).

The new LV will use the latest PV count of the corresponding VG. We will check the latest PV count of the VG when we create the LV. For the above example, the strip number will be three after extending the VG.